planner/types.go

```diff
@@ -18,11 +18,41 @@ type PlanNode interface {
 
 	// Helper methods for manipulating a plan
 	// These methods are used during planning
-	AddSuccessor(PlanNode)
-	AddPredecessor(PlanNode)
+	AddSuccessors(...PlanNode)
+	AddPredecessors(...PlanNode)
 	RemovePredecessor(PlanNode)
 }
 
+// QueryPlan holds the roots of the query plan DAG.
+// Roots correspond to nodes that produce final results.
+type QueryPlan struct {
+	roots map[PlanNode]struct{}
```

Comment:
I've never seen this before.  Is this the idiomatic way to have something like `std::unordered_set` in go?

Reply:
I'm not sure. I've seen it done `map[PlanNode]bool` too, but `struct{}` has a size of 0.